### ISSUE TITLE
Feature request: Option to disable autocommit

### ISSUE THREAD

**Issue by tedstriker2** [pawelsalawa/sqlitestudio#4510]
### Details
I just made a huge mistake with an update query - I left out the where clause, and updated every row. I'm much more used to Oracle, which doesn't autocommit. Luckily I had a backup of my database from yesterday.

I'm not seeing an option in SQLiteStudio to change the autocommit behavior, and think it would be great to have that.

**Comment by pawelsalawa** [comment]
I spent quite a bit of time exploring different approaches to support non-autocommit workflows in Letos, but every solution kept running into serious architectural and UX problems.

For example, using a separate SQLite connection per SQL editor tab initially seems like the obvious approach, but in practice the application often needs to synchronize internal database state between connections so that other features continue to work correctly. While technically possible, it introduces a large surface for subtle bugs, including potentially dangerous data consistency issues.

There is also the locking aspect. SQLite itself allows only a single write transaction at a time, so independent editor windows would frequently block each other anyway. A long-running transaction left open in one SQL editor could even block operations that may appear completely unrelated - for example simply reading data somewhere else, either in another part of Letos or even from a different application accessing the same database file. WAL mode helps mitigate this, but not every user knows about it, and not everyone wants to use WAL mode in the first place. Realistically, this approach would be far more viable if SQLite supported multiple concurrent write transactions without effectively locking the whole database for other operations.

I also experimented with keeping a single shared connection and managing transactions globally, but then every other database-related operation and window in the application has to become transaction-aware as well. The UI/UX complexity also grows very quickly - it becomes much harder to keep the behavior intuitive and unambiguous for users.

One of the main goals of the 4.x series was simplifying and cleaning up the interface, so after weighing the complexity, maintenance cost, and risk of introducing hard-to-debug issues, I decided not to implement transaction handling in the same style as some other database IDEs.

That said, I do agree the accidental UPDATE/DELETE case is important, and I’m still thinking about alternative ways to help protect against these kinds of mistakes.

**Comment by tedstriker2** [comment]
That sounds wise, but do those same issues present themselves if a user types "BEGIN;" in every new window?

**Comment by pawelsalawa** [comment]
Yes - executing `BEGIN` manually has essentially the same underlying consequences as the "global transaction" approach I mentioned earlier. From that point on, all changes become part of the same transaction, including schema modifications, index creation, etc. Depending on the locking mode and journal mode, it may also lock access to the database file for other operations.

The important difference is that manually executing `BEGIN` does not imply full application-level transaction awareness. In other words, the application does not try to automatically coordinate every UI component and internal cache around that transaction state. Users who type `BEGIN` themselves are usually much more aware of the implications and limitations than users who simply click a Commit/Rollback button in the UI and expect everything to behave transparently.

From experience, I know that if I expose transaction controls directly in the UI, users will naturally expect the application to fully handle all edge cases and synchronization automatically. That would likely result in a large number of bug reports and confusing UI situations caused by SQLite’s transactional and locking model rather than by the application itself.

If SQLite supported multiple concurrent write/read transactions more naturally - without effectively blocking database file for other operations - this would be much easier to design cleanly. In that world, having a dedicated connection and transaction scope per SQL editor window would actually make a lot of sense and would probably feel intuitive.

At a high level the feature sounds straightforward, and initially I also thought it should be relatively simple. The complexity only becomes visible once you start dealing with all the interaction details between editor windows, schema refreshes, internal caches, object trees, query results, locking behavior, and user expectations.

So I’m not abandoning the idea entirely. I just need to find an approach that improves safety without introducing confusing UI/UX behavior or fragile architecture.

**Comment by pawelsalawa** [comment]
I've created #5670 to track alternative solution. Initially I scheduled it for 4.1.0, but if I figure it out early enough, it may appear even in 4.0.0.
